tensorflow_datasets/core/custom_dataset/image_folder.py

```diff
@@ -0,0 +1,176 @@
+        xxz.png
+      label2/
+        xxx.png
+        xxy.png
+        xxz.png
+    split_name/  # Ex: 'test'
+      ...
+  ```
+
+  To use it:
+
+  ```
+  builder = tfds.ImageFolder(root_dir='path/to/manual_dir/')
+  print(builder.info)  # Splits, num examples,... are automatically calculated
+  ds = builder.as_dataset(split='split_name', shuffle_files=True)
+  tfds.show_examples(ds, builder.info)
+  ```
+
+  """
+
+  VERSION = core.Version('2.0.0')
+
+  def __init__(self, root_dir):
+    root_dir = os.path.expanduser(root_dir)
+    super(ImageFolder, self).__init__(
+        data_dir=root_dir, version=str(self.VERSION), config=None)
+
+    # Reset `_data_dir` as it should not change to DATA_DIR/Version
+    self._data_dir = root_dir
+
+    # dict[split_name][label_name] = list(img_paths)
+    self._split_label_images, labels = _get_split_label_images(root_dir)
+
+    # Calculate and update DatasetInfo labels
+    self.info.features['label'].names = sorted(list(labels))
+
+    # Calculate and update DatasetInfo splits
+    splits = {
+        split_name: sum(len(l)
+                        for l in self._split_label_images[split_name].values())
+        for split_name in self._split_label_images
+    }
+    split_dict = core.SplitDict(self.name)
+    for split_name, size in splits.items():
+      split_dict.add(
+          core.SplitInfo(name=split_name, shard_lengths=[size]))
+    self.info.update_splits_if_different(split_dict)
+
+  def _info(self):
+    return core.DatasetInfo(
+        builder=self,
+        description='Generic image classification dataset.',
+        features=core.features.FeaturesDict({
+            'image': core.features.Image(),
+            'label': core.features.ClassLabel(),
+            'image/filename': core.features.Text(),
+        }),
+        supervised_keys=('image', 'label'),
+    )
+
+  def _download_and_prepare(self, **kwargs):
+    raise NotImplementedError('No need to call download_and_prepare function.')
+
+  def download_and_prepare(self, **kwargs):
+    raise NotImplementedError('No need to call download_and_prepare function.')
+
+  def _process_ds(self, path, label):
+    image_path = tf.strings.split(path, os.path.sep)[-3:]
+    img = tf.io.read_file(path)
+    img = tf.image.decode_jpeg(img, channels=3)
+    img = tf.image.convert_image_dtype(img, tf.uint8)
+    return {
+        'image': img,
+        'label': tf.cast(label, tf.int64),
+        'image/filename': tf.strings.reduce_join(image_path)
+    }
+
+  def _as_dataset(
+      self,
+      split,
+      shuffle_files=False,
+      decoders=None,
+      read_config=None):
+    """Generate dataset for given split"""
+    del decoders
+    del read_config
+    if split not in self.info.splits.keys():
+      raise ValueError('Split name {} not present in {}'
+                       .format(split, self._split_label_images.keys()))
+
+    img_paths = sorted(
+        img for l in self._split_label_images[split].values() for img in l)
```

Comment:
You could use `_split_label_images[split].items()` to extract both labels & images at once.

Then use `images, labels = zip(*inputs)` to unpack them